dna/models/torch_modules/attention_mlp.py

```diff
@@ -0,0 +1,68 @@
+import torch
+from torch import nn
+from torch_transformer import Encoder
+
+from . import PyTorchRandomStateContext
+from . import F_ACTIVATIONS
+from .submodule import Submodule
+from dna.utils import get_reduction_function
+
+
+class AttentionMLP(nn.Module):
+    """
+    The Attention MLP applies attention to a sequence without considering the ordering of the inputs in the sequence.
+    It then applies a reduction to the sequence to produce an encoded sequence as a single vector. This vector is
+    concatenated to another features vector and then passed through an MLP.
+    """
+
+    def __init__(
+        self, n_layers: int, n_heads: int, in_features: int, attention_in_features: int, attention_hidden_features,
+        attention_activation_name: str, dropout: float, reduction: str, mlp_extra_input_size: int,
+        mlp_hidden_layer_size: int, mlp_n_hidden_layers: int, mlp_activation_name: str, output_size: int,
+        mlp_use_batch_norm: bool, mlp_use_skip: bool, *, device: str, seed: int
+    ):
+
+        super().__init__()
+
+        if attention_in_features % n_heads != 0:
+            raise ValueError(
+                'attention_in_features must be divisible by n_heads. {0} is not divisible by {1}'.format(
+                    attention_in_features, n_heads
+                )
+            )
+
+        # The embedder maps one hot encoded inputs to a feature space with a dimension size that the encoder can handle
+        # Without the embedder, the encoder input feature size is restricted to the number of one hot encodings
+        # This would be problematic because attention_in_features must be divisible by n_heads
```

Comment:
Why does the input size to the attention module need to be divisible by n_heads? Is the input vector being split into n_heads pieces? If so, we should probably be copying the input vector n_heads times, not sure though.

Reply:
I'll do some digging and get more information on this.

Reply:
There is a helper function in the MultiHeadedAttention module which is used by the Encoder. It's called _reshape_to_batches. This function changes the shape of the input tensor such that the batch size is multiplied by the number of heads and the input feature size is divided by the number of heads. The reason why the input feature size must be divisible by the number of heads is for this resize to work.

Here's an example of output from the python console where x is a potential input tensor with a batch size of two, sequence length of 3, and input feature size of 4. The number of heads in this example would be 2, where 4 is divisible by 2.
x.shape
torch.Size([2, 3, 4])
x
tensor([[[ 1,  2,  3,  4],
         [ 5,  6,  7,  8],
         [ 9, 10, 11, 12]],
        [[13, 14, 15, 16],
         [17, 18, 19, 20],
         [21, 22, 23, 24]]])
self._reshape_to_batches(x).shape
torch.Size([4, 3, 2])
self._reshape_to_batches(x)
tensor([[[ 1,  2],
         [ 5,  6],
         [ 9, 10]],
        [[ 3,  4],
         [ 7,  8],
         [11, 12]],
        [[13, 14],
         [17, 18],
         [21, 22]],
        [[15, 16],
         [19, 20],
         [23, 24]]])

Reply:
What this does is it splits the input feature vectors into a number of heads and then performs matrix operations on the heads. This demonstrates that if the Encoder were to receive one hot encoded input vectors, all but one of the heads would be all 0s. The remaining head would have a single 1 in it. Therefor the first layer in the Encoder would be getting little to no information. Another reason to embed the one hot encodings first.

Reply:
There is no need to copy the input vectors n_head times because we can just make sure that the embedding mlp has the correct number of output features.

Reply:
> What this does is it splits the input feature vectors into a number of heads and then performs matrix operations on the heads.

I understand that mechanically the input size would need to be divisible by the number of heads for this work. Why do it this way? It seems that each attention head would then only be able to attend to a small part of the input, not the whole sequence.

Reply:
I was following this conversation and realized that I wasn't sure why they do that either.  It turns out that was something I missed in the paper: https://www.quora.com/How-does-the-multi-head-attention-mechanism-work-in-deep-learning